### PR TITLE
WIP: Change the way `xclip` handles clearing

### PR DESCRIPTION
**Pull request by jacob-horton**
Only clear the `xclip` clipboard if nothing has been copied since the password. This prevents something else being cleared from the clipboard if it was copied between copying the password and it being cleared.

This does, however, mean that the password is stored in memory (in the `previously_copied` variable) until it is cleared from the clipboard. Also, if `--clear-after` is not used, it will stay in the `previously_copied` variable until overwritten by the next copy. This is potentially a security risk, so I am unsure whether this feature is worth the reduction in security.

### CODE REVIEW

**Comment by fdw** [comment]
Hey, sorry this took so long. I wanted to revisit it with a fresh mind, but it fell off my radar 🙁

However, it doesn't add a lot of complexity and it might help at least one person (you 😉), so I'll leave a short code review and then we can merge it.

Thanks for your patience!

File: src/rofi_rbw/clipboarder.py [hunk 1]
```diff
@@ -77,10 +77,27 @@ def copy_to_clipboard(self, characters: str) -> None:
             encoding='utf-8'
         )
 
+        self.previously_copied = characters
+
+    def get_clipboard(self) -> str:
```
Comment (fdw) [inline src/rofi_rbw/clipboarder.py:82]: Can we please rename that to something better fitting? It doesn't get the clipboard, it's about the clipboard's content. So maybe `fetch_content` or `fetch_clipboard_content`?

File: src/rofi_rbw/clipboarder.py [hunk 2]
```diff
@@ -77,10 +77,27 @@ def copy_to_clipboard(self, characters: str) -> None:
             encoding='utf-8'
         )
 
+        self.previously_copied = characters
```
Comment (fdw) [inline src/rofi_rbw/clipboarder.py:80]: I'd like to declare this with the class definition, like
```
class XClipClipboarder(Clipboarder):
    __previously_copied: str
```

Also, I don't find the name ideal: At this moment, it's not "previously" copied, but currently. Maybe `last_copied_characters`? Or do you have a better idea?

**Review by fdw — CHANGES_REQUESTED** [review]

File: src/rofi_rbw/clipboarder.py (hunk 1 above)
Comment (jacob-horton) [reply src/rofi_rbw/clipboarder.py]: Sounds good, I changed it to `fetch_clipboard_content`

File: src/rofi_rbw/clipboarder.py (hunk 2 above)
Comment (jacob-horton) [reply src/rofi_rbw/clipboarder.py]: I think `last_copied_characters` is good, I've now changed it. Sorry I took a while to get back to this

**Comment by jacob-horton** [comment]
I have just found that the same "issue" occurs with `xsel`, I don't know if there has been a change, or if I missed it before. Would you like me to do the same changes for that? And if so, should I add it to the base class? I don't use Wayland, so I do not know if `wtype` has this issue

**Comment by fdw** [comment]
Feel free to add it to `xsel`, too 🙂

However, I wouldn't want to do it in the base class, as not every clipboarder necessarily has that problem.

**Comment by jacob-horton** [comment]
I've now added it for `xsel` :slightly_smiling_face:

File: src/rofi_rbw/clipboarder.py [hunk 3]
```diff
@@ -49,17 +49,35 @@ def copy_to_clipboard(self, characters: str) -> None:
...
+        self.__last_copied_characters = characters
+
+    def fetch_clipboard_content(self) -> str:
+        return run([
+            'xsel',
+            '--output',
+            '--clipboard',
+        ],
+            capture_output=True,
+            text=True
```
Comment (fdw) [inline src/rofi_rbw/clipboarder.py:61]: The other calls use `encoding='utf-8'`. Is there an advantage to this?

**Comment by fdw** [comment]
Thank you! 🙂

**Comment by fdw** [comment]
@jacob-horton Can you please have a look at my last question and maybe rebase again?

File: src/rofi_rbw/clipboarder.py (hunk 3 above)
Comment (jacob-horton) [reply src/rofi_rbw/clipboarder.py:61]: Sorry, not that I am aware of. I have just previously used `text=True`. I have just tried it with `encoding='utf-8'` and it seems to be working as intended. Would you like me to push this change?

Comment (fdw) [reply src/rofi_rbw/clipboarder.py:61]: Yep, that would be nice for consistency. Thanks :)

**Comment by fdw** [comment]
I had some problems merging it directly, so I rebased, squashed and merged it manually. Hope this is okay for you 🙂

**Pull request closed without merging**